### PR TITLE
Fix for issue #23

### LINKED ISSUE #23: YamlEditor appendToList and insertIntoList inserts new item into next yaml item rather than at end of list
The following results in an Unhandled Exception "Assertion failed: (package:yaml_edit) Modification did not result in expected result."

```
import 'dart:io';

import 'package:yaml_edit/yaml_edit.dart';

Future<void> main() async {
  YamlEditor yamlEditorFile =
      YamlEditor(File('test/gh_test.yml').readAsStringSync());

  try {
    // yamlEditorFile.insertIntoList(
    //     ['school'],
    //     3,
    //     {
    //       'student': {'first_name': 'Henry', 'last_name': 'Smith'}
    //     });

    yamlEditorFile.appendToList([
      'school'
    ], {
      'student': {'first_name': 'Henry', 'last_name': 'Smith'}
    });
  } on Exception catch (e) {
    print(e);
  }

  print(yamlEditorFile);
}

```
gh_test.yml is: 

```
# comment
school:
  - student:
      first_name: Harry
      last_name: "Reynolds"
  - student:
      first_name: "Sally"
      last_name: Rheims
  - student:
      first_name: "George"
      last_name: "Silver"
teachers:
  - english:
      name: "Johnson"
  - science:
      name: "Smith"
```

The insertIntoList call results in the same if the index is the next one (i.e., 3 - commented code in example above). 

Expected behavior in all cases is that it inserts it correctly in the school list after student George Silver.

Instead the result in both examples is that the inserted list item (Henry Smith) is inserted incorrectly as the first item of the next list (i.e., teachers) as shown by the "Modification did not result in expected result" message.":

```

> teachers:
>   - student:
>       first_name: Henry
>       last_name: Smith
>   - english:
>       name: "Johnson"
>   - science:
>       name: "Smith"
```
the full error message: 
```

Unhandled exception:
Assertion failed: (package:yaml_edit) Modification did not result in expected result.

# YAML before edit:
> # comment
> school:
>   - student:
>       first_name: Harry
>       last_name: "Reynolds"
>   - student:
>       first_name: "Sally"
>       last_name: Rheims
>   - student:
>       first_name: "George"
>       last_name: "Silver"
> teachers:
>   - english:
>       name: "Johnson"
>   - science:
>       name: "Smith"

# YAML after edit:
> # comment
> school:
>   - student:

>       first_name: Harry
>       last_name: "Reynolds"
>   - student:
>       first_name: "Sally"
>       last_name: Rheims
>   - student:
>       first_name: "George"
>       last_name: "Silver"
> teachers:
>   - student:
>       first_name: Henry
>       last_name: Smith
>   - english:
>       name: "Johnson"
>   - science:
>       name: "Smith"

Please file an issue at:
https://github.com/google/dart-neats/issues/new?labels=pkg%3Ayaml_edit%2C+pending-triage&template=yaml_edit.md

#0      YamlEditor._performEdit
#1      YamlEditor.insertIntoList
#2      YamlEditor.appendToList
#3      main

#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)

Exited (255)

```
So it appears it can't detect the end of the list.  The example yaml is correctly validated. If I insert a non-list in between the two lists in the example yaml file, for example the boolean "public: false", I get "Mapping Values are not allowed here" as it attempts to insert it under the boolean yaml:

Here's that error message:
```

Error on line 13, column 12: Mapping values are not allowed here. Did you miss a colon earlier?
   ╷
13 │   - student:
   │            ^
   ╵
# comment
school:
  - student:
      first_name: Harry
      last_name: "Reynolds"
  - student:
      first_name: "Sally"
      last_name: Rheims
  - student:
      first_name: "George"
      last_name: "Silver"
public: false
  - student:
      first_name: Henry
      last_name: Smith
teachers:
  - english:
      name: "Johnson"
  - science:
      name: "Smith"
Exited

```

### PR DESCRIPTION
**Pull request by Ishad-M-I-M**
The `nextNewLineIndex` is calculated incorrectly in `_appendToBlockList` function
Fixes #23

### CODE REVIEW

**Review by jonasfj — CHANGES_REQUESTED** [review]
We definitely need a test case that proves this works.

**Comment by jonasfj** [comment]
Also, if you write:

Fixes #23

In the pull request description, then it'll be linked to the issue, and the issue will be closed when this pull-request is landed.


--------------------

It might also be nice to make a few more tests to help demonstrates that we're not introducing any new issues with this change :D

There most likely is variations of list append we haven't got good tests cases for.

**Comment by jonasfj** [comment]
@Ishad-M-I-M if I forget, feel free to @jonasfj ping me; I'd love to get this landed. Once test pass and ideally we have few more test cases.

**Comment by Ishad-M-I-M** [comment]
@jonasfj added one testcase for each `appendToList`  and `insertToList`.

File: lib/src/list_mutations.dart [hunk 1]
```diff
@@ -117,7 +117,7 @@ SourceEdit _appendToBlockList(
   // Adjusts offset to after the trailing newline of the last entry, if it exists
   if (list.isNotEmpty) {
     final lastValueSpanEnd = list.nodes.last.span.end.offset;
-    final nextNewLineIndex = yaml.indexOf('\n', lastValueSpanEnd);
+    final nextNewLineIndex = yaml.indexOf('\n', lastValueSpanEnd-1);
```
Comment (jonasfj) [inline lib/src/list_mutations.dart:120]: make sure your editor runs `dart format` on save, it'll make your life a lot easier :D

File: test/insert_test.dart [hunk 2]
```diff
@@ -98,6 +98,43 @@ void main() {
...
+  - z:
+      x: 5
+      y: 6
+b:
+  - w:
+      m: 2
+      n: 4
+'''));
+    });
+
```
Comment (jonasfj) [inline test/insert_test.dart:137]: May I propose an additional test along the lines of:

```dart
for (var i = 0; i < 3; i++) {
  test('block insert at $i', () {
    final yamlEditor = YamlEditor('''
a:
  - z:
      x: 1
      y: 2
  - z:
      x: 3
      y: 4
b: "next line"
''');
    // Check that this doesn't throw an internal error.
    yamlEditor.insertIntoList(['a'], i, {'z' : {'x': 5, 'y': 6}});
  });
}
```

If you look at the original issues you'll see that `insertIntoList` can throw an exception, if modification results in invalid YAML.
Specifically, it'll throw:
> Assertion failed: (package:yaml_edit) Modification did not result in expected result.

So while it's wonderful to have some tests where we check that the output is as expected, it's also really nice to have a lot of tests that tries different input without making assertions on the output. Such tests are easy to write a lot of, and they check that the `Assertion failed` error isn't thrown. So while it's not a perfect test, it's a test that won't break if we change the output of formatting a little -- that kind of test is easy to maintain.

**Review by jonasfj — CHANGES_REQUESTED** [review]

File: test/insert_test.dart (hunk 2 above)
Comment (Ishad-M-I-M) [reply test/insert_test.dart]: Added few more test cases considering different patterns.

**Review by jonasfj — APPROVED** [review]

**Pull request merged**